src/blockchain.rs

```diff
@@ -217,86 +219,131 @@ impl Blockchain {
             return false;
         }
 
         let transactions_valid = block
             .transactions()
             .iter()
-            .all(|tx| self.validate_transaction(block, tx, fork_tuple));
+            .all(|tx| self.validate_transaction(previous_block, block, tx, fork_tuple));
 
         transactions_valid
     }
 
     fn validate_transaction(
         &self,
+        previous_block: &Block,
         block: &Block,
         tx: &Transaction,
         fork_tuple: &ForkTuple,
     ) -> bool {
-        // check that our sigs are valid
-        if let Some(output_slip) = self.utxoset.output_slip_from_slip_id(&tx.core.inputs()[0]) {
-            // let serialized_tx: Vec<u8> = tx.core.clone().into();
-            // make_message_from_bytes(&serialized_tx[..]);
-            let hash_message = tx.core.hash();
-            if !verify_message_signature(&hash_message, &tx.signature(), &(output_slip.address())) {
-                return false;
-            };
-
-            // validate our slips
-            let inputs_are_valid = tx.core.inputs().iter().all(|input| {
-                return !self
-                    .utxoset
-                    .is_slip_spent_at_block(input, block, fork_tuple);
-            });
-
-            if !inputs_are_valid {
-                return false;
-            }
+        match tx.core.broadcast_type() {
+            TransactionType::Normal => {
+                if tx.core.inputs().len() == 0 && tx.core.outputs().len() == 0 {
+                    return true;
+                }
 
-            // valuidate that inputs are unspent
-            let input_amt: u64 = tx
-                .core
-                .inputs()
-                .iter()
-                .map(|input| {
-                    self.utxoset
-                        .output_slip_from_slip_id(input)
-                        .unwrap()
-                        .amount()
-                })
-                .sum();
-
-            let output_amt: u64 = tx.core.outputs().iter().map(|output| output.amount()).sum();
-
-            if input_amt != output_amt {
-                return false;
+                if let Some(output_slip) =
+                    self.utxoset.output_slip_from_slip_id(&tx.core.inputs()[0])
+                {
+                    let hash_message = tx.core.hash();
+                    if !verify_message_signature(
+                        &hash_message,
+                        &tx.signature(),
+                        &(output_slip.address()),
+                    ) {
+                        println!("SIGNATURE IS NOT VALID");
+                        return false;
+                    };
+
+                    // validate our slips
+                    let inputs_are_valid = tx.core.inputs().iter().all(|input| {
+                        return !self
+                            .utxoset
+                            .is_slip_spent_at_block(input, block, fork_tuple);
+                    });
```

Comment:
hmmm, there must be a nicer way to do this... 

Can we match on this and reduce it to just sticking a "?" on the end and get rid of inputs_are_valid?

Happy to accept this as-is, but maybe a little "TODO" would be nice to acknowledge that this might not be the rustiest way to do it.